src/pure/List.mo

```diff
@@ -11,147 +20,778 @@ module {
+    };
+
+  /// Create a new list with only those elements of the original list for which
+  /// the given function (often called the _predicate_) returns true.
+  ///
+  /// Example:
+  /// ```motoko include=initialize
+  /// List.filter<Nat>(?(0, ?(1, ?(2, null))), func n = n != 1) // => ?(0, ?(2, null))
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  public func filter<T>(list : List<T>, f : T -> Bool) : List<T> =
+    switch list {
+      case null null;
+      case (?(h, t)) if (f h) ?(h, filter(t, f)) else filter(t, f)
+    };
+
+  /// Call the given function on each list element, and collect the non-null results
+  /// in a new list.
+  ///
+  /// Example:
+  /// ```motoko include=initialize
+  /// List.filterMap<Nat, Nat>(
+  ///   ?(1, ?(2, ?(3, null))),
+  ///   func n = if (n > 1) ?(n * 2) else null
+  /// ) // => ?(4, ?(6, null))
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `f` runs in O(1) time and space.
+  public func filterMap<T, R>(list : List<T>, f : T -> ?R) : List<R> =
+    switch list {
+      case null null;
+      case (?(h, t)) {
+        let ?v = f h else return filterMap(t, f);
+        ?(v, filterMap(t, f))
+      }
+    };
+
+  /// Maps a `Result`-returning function `f` over a `List` and returns either
+  /// the first error or a list of successful values.
+  ///
+  /// Example:
+  /// ```motoko include=initialize
+  /// List.mapResult<Nat, Nat, Text>(
+  ///   ?(1, ?(2, ?(3, null))),
+  ///   func n = if (n > 0) #ok(n * 2) else #err "Some element is zero"
+  /// ) // => #ok ?(2, ?(4, ?(6, null))
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `f` runs in O(1) time and space.
+  public func mapResult<T, R, E>(list : List<T>, f : T -> Result.Result<R, E>) : Result.Result<List<R>, E> =
+    switch list {
+      case null #ok null;
+      case (?(h, t)) {
+        switch (f h, mapResult(t, f)) {
+          case (#ok r, #ok l) #ok (?(r, l));
+          case (#err e, _) #err e;
+          case (_, #err e) #err e
+        }
+      }
+    };
+
+  /// Create two new lists from the results of a given function (`f`).
+  /// The first list only includes the elements for which the given
+  /// function `f` returns true and the second list only includes
+  /// the elements for which the function returns false.
+  ///
+  /// Example:
+  /// ```motoko include=initialize
+  /// List.partition<Nat>(?(0, ?(1, ?(2, null))), func n = n != 1) // => (?(0, ?(2, null)), ?(1, null))
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `f` runs in O(1) time and space.
+  public func partition<T>(list : List<T>, f : T -> Bool) : (List<T>, List<T>) =
+    switch list {
+      case null (null, null);
+      case (?(h, t)) {
+        let left = f h;
+        let (l, r) = partition(t, f);
```

Comment:
Could be a problem with stack overflow for longer list. We had such issues in the past. I suggest using a loop.